morango/models/core.py

```diff
@@ -115,49 +113,55 @@ class InstanceIDModel(UUIDModelMixin):
     system_id = models.CharField(max_length=100, blank=True)
 
     @classmethod
+    @transaction.atomic
     def get_or_create_current_instance(cls):
         """Get the instance model corresponding to the current system, or create a new
         one if the system is new or its properties have changed (e.g. OS from upgrade)."""
 
-        # on Android, platform.platform() barfs, so we handle that safely here
+        # check if a matching legacy instance ID is already current, and don't mess with it
+        kwargs = get_0_4_system_parameters(
+            database_id=DatabaseIDModel.get_or_create_current_database_id().id
+        )
         try:
-            plat = platform.platform()
-        except:  # noqa: E722
-            plat = "Unknown (Android?)"
-
-        kwargs = {
-            "platform": plat,
-            "hostname": platform.node(),
-            "sysversion": sys.version,
-            "database": DatabaseIDModel.get_or_create_current_database_id(),
-            "db_path": os.path.abspath(settings.DATABASES["default"]["NAME"]),
-            "system_id": os.environ.get("MORANGO_SYSTEM_ID", ""),
-        }
-
-        # try to get the MAC address, but exclude it if it was a fake (random) address
-        mac = uuid.getnode()
-        if (mac >> 40) % 2 == 0:  # 8th bit (of 48 bits, from left) is 1 if MAC is fake
-            hashable_identifier = "{}:{}".format(kwargs["database"].id, mac)
-            kwargs["node_id"] = hashlib.sha1(
-                hashable_identifier.encode("utf-8")
-            ).hexdigest()[:20]
-        else:
-            kwargs["node_id"] = ""
+            obj = InstanceIDModel.objects.get(current=True, **kwargs)
+            return obj, False
+        except InstanceIDModel.DoesNotExist:
+            pass
+
+        # calculate the new ID based on system ID and mac address
+        kwargs["system_id"] = get_0_5_system_id()
+        kwargs["node_id"] = get_0_5_mac_address()
+        kwargs["id"] = sha2_uuid(
+            kwargs["database_id"], kwargs["system_id"], kwargs["node_id"]
+        )
+        kwargs["current"] = True
 
-        # do within transaction so we only ever have 1 current instance ID
-        with transaction.atomic():
-            InstanceIDModel.objects.filter(current=True).update(current=False)
-            obj, created = InstanceIDModel.objects.get_or_create(**kwargs)
-            obj.current = True
+        # ensure we only ever have 1 current instance ID
+        InstanceIDModel.objects.exclude(id=kwargs["id"], current=False).update(
+            current=False
+        )
+        # create the model, or get existing if one already exists with this ID
+        obj, created = InstanceIDModel.objects.get_or_create(
+            id=kwargs["id"], defaults=kwargs
+        )
+        # update any of the attributes that have changed
+        changed = False
+        for key, val in kwargs.items():
+            if getattr(obj, key) != val:
+                setattr(obj, key, val)
```

Comment:
Would using `update_or_create` instead of `get_or_create` work the same?

Reply:
Yes, I think so! Had forgotten about that one.